### PR TITLE
[RFC] feat: markdown compatible tables

### PR DESCRIPTION
**Pull request by sakulstra**
## Motivation

When using foundry features(inspect for storage layout, but potentially gas report etc), it often happens that in the end i want to put these results into a document. What i noticed ppl tend to do is: making screenshots of the terminal and embedding it into markdown. This imo is not so good as it makes this stuff hard to maintain and a bit cumbersome. Therefore i'd suggest to allow generating markdown compatible tables.

## Solution

The `comfy_table` crate used inside foundry provides a markdown preset that can be used.

I think is worth discussing:
Does this need a flag or should it just be the default?

I don't really see the value in "non markdown" tables (as markdown is perfectly readable in console as well).
If there's no opposition, would delete the flag and just use this preset in `all` tables.

Before:
+----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------+
| Name                             | Type                                                                                   | Slot | Offset | Bytes | Contract                                                                                                          |
+=======================================================================================================================================================================================================================================================================+
| _balances                        | mapping(address => uint256)                                                            | 0    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _allowances                      | mapping(address => mapping(address => uint256))                                        | 1    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _totalSupply                     | uint256                                                                                | 2    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _name                            | string                                                                                 | 3    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _symbol                          | string                                                                                 | 4    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _decimals                        | uint8                                                                                  | 5    | 0      | 1     | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _votingSnapshots                 | mapping(address => mapping(uint256 => struct GovernancePowerDelegationERC20.Snapshot)) | 6    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _votingSnapshotsCounts           | mapping(address => uint256)                                                            | 7    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _aaveGovernance                  | contract ITransferHook                                                                 | 8    | 0      | 20    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| lastInitializedRevision          | uint256                                                                                | 9    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| ______gap                        | uint256[50]                                                                            | 10   | 0      | 1600  | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| assets                           | mapping(address => struct AaveDistributionManager.AssetData)                           | 60   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| stakerRewardsToClaim             | mapping(address => uint256)                                                            | 61   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| stakersCooldowns                 | mapping(address => uint256)                                                            | 62   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _votingDelegates                 | mapping(address => address)                                                            | 63   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _propositionPowerSnapshots       | mapping(address => mapping(uint256 => struct GovernancePowerDelegationERC20.Snapshot)) | 64   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _propositionPowerSnapshotsCounts | mapping(address => uint256)                                                            | 65   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _propositionPowerDelegates       | mapping(address => address)                                                            | 66   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| DOMAIN_SEPARATOR                 | bytes32                                                                                | 67   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
|----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------|
| _nonces                          | mapping(address => uint256)                                                            | 68   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
+----------------------------------+----------------------------------------------------------------------------------------+------+--------+-------+-------------------------------------------------------------------------------------------------------------------+

After:
| Name                             | Type                                                                                   | Slot | Offset | Bytes | Contract                                                                                                          |
|----------------------------------|----------------------------------------------------------------------------------------|------|--------|-------|-------------------------------------------------------------------------------------------------------------------|
| _balances                        | mapping(address => uint256)                                                            | 0    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _allowances                      | mapping(address => mapping(address => uint256))                                        | 1    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _totalSupply                     | uint256                                                                                | 2    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _name                            | string                                                                                 | 3    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _symbol                          | string                                                                                 | 4    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _decimals                        | uint8                                                                                  | 5    | 0      | 1     | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _votingSnapshots                 | mapping(address => mapping(uint256 => struct GovernancePowerDelegationERC20.Snapshot)) | 6    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _votingSnapshotsCounts           | mapping(address => uint256)                                                            | 7    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _aaveGovernance                  | contract ITransferHook                                                                 | 8    | 0      | 20    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| lastInitializedRevision          | uint256                                                                                | 9    | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| ______gap                        | uint256[50]                                                                            | 10   | 0      | 1600  | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| assets                           | mapping(address => struct AaveDistributionManager.AssetData)                           | 60   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| stakerRewardsToClaim             | mapping(address => uint256)                                                            | 61   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| stakersCooldowns                 | mapping(address => uint256)                                                            | 62   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _votingDelegates                 | mapping(address => address)                                                            | 63   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _propositionPowerSnapshots       | mapping(address => mapping(uint256 => struct GovernancePowerDelegationERC20.Snapshot)) | 64   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _propositionPowerSnapshotsCounts | mapping(address => uint256)                                                            | 65   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _propositionPowerDelegates       | mapping(address => address)                                                            | 66   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| DOMAIN_SEPARATOR                 | bytes32                                                                                | 67   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |
| _nonces                          | mapping(address => uint256)                                                            | 68   | 0      | 32    | src/etherscan/mainnet_0xe42f02713aec989132c1755117f768dbea523d2f/StakedTokenV2Rev3/Contract.sol:StakedTokenV2Rev3 |

### CODE REVIEW

**Comment by mds1** [comment]
> it often happens that in the end i want to put these results into a document. What i noticed ppl tend to do is: making screenshots of the terminal and embedding it into markdown. This imo is not so good as it makes this stuff hard to maintain and a bit cumbersome. Therefore i'd suggest to allow generating markdown compatible tables.

This makes sense to me, I like this idea

> Does this need a flag or should it just be the default? I don't really see the value in "non markdown" tables (as markdown is perfectly readable in console as well). If there's no opposition, would delete the flag and just use this preset in all tables.

Do you mind posting a screenshot of what the terminal output looks like for markdown vs. non-markdown tables? As long as both versions look equally readable it seems fine to make markdown the default to avoid needing to introduce a flag

File: forge/src/gas_report.rs
Comment (sakulstra) [inline forge/src/gas_report.rs:162]: newline needed here as otherwise gas reports look a bit wrecked as there's no longer a round border on top & bottom

**Comment by mds1** [comment]
Thanks for the screenshots! I do think for e.g. contracts with a lot of storage the non-markdown version with extra lines is easier to read across a row, OTOH I can double click some text to highlight that whole row to make sure I don't misread something. So I think in an effort to keep the API simpler I'd be ok with markdown as the default and removing the verbose approach unless someone requests having it back afterwards

Will defer to @mattsse from here to review/merge

**Review by mattsse — APPROVED** [review]
lgtm,

good to merge @mds1 ?

**Comment by mds1** [comment]
Tested quickly, lgtm pending the rebase + CI passing.

I do think it's a but uglier now since there's no top/bottom lines on the table, but unfortunately adding those isn't markdown compatible. Maybe a blank line before the table would be nice (see below), but we can iterate and improve the style in future PRs

```
forge build --sizes
[⠆] Compiling...
No files changed, compilation skipped
| Contract                | Size (kB) | Margin (kB) |
|-------------------------|-----------|-------------|
(truncated the rest for clarity, right above the table
header a line break may be nice)
```

**Pull request merged**
